test/list_test.go

```diff
@@ -178,32 +178,100 @@ func TestList(t *testing.T) {
 	}
 }
 
-// BenchmarkList is a benchmark for the Get operation.
 func BenchmarkList(b *testing.B) {
-	for _, backendType := range []string{endpoint.SQLiteBackend, endpoint.DQLiteBackend} {
-		b.Run(backendType, func(b *testing.B) {
-			b.StopTimer()
-			g := NewWithT(b)
+	setup := func(ctx context.Context, tx *sql.Tx, payloadSize, n int) error {
+		if err := insertMany(ctx, tx, "key", payloadSize, n); err != nil {
+			return err
+		}
+		b.Log("insert", n)
+
+		if err := updateMany(ctx, tx, "key", payloadSize, n/2); err != nil {
+			return err
+		}
+		b.Log("update", n)
+
+		if err := deleteMany(ctx, tx, "key", n/2); err != nil {
+			return err
+		}
+		b.Log("delete", n)
+		return nil
+	}
+	backends := []string{endpoint.SQLiteBackend, endpoint.DQLiteBackend}
+	for _, backendType := range backends {
+		payloads := []struct {
+			name string
+			size int
+		}{{
+			name: "tiny",
+			size: 100,
+		}, {
+			name: "page-fitting",
+			size: 1000,
+		}, {
+			name: "page-overflowing",
+			size: 5000,
+		}}
+		for _, payload := range payloads {
+			b.Run(fmt.Sprintf("%s-%s", backendType, payload.name), func(b *testing.B) {
+				b.Run("all", func(b *testing.B) {
+					b.StopTimer()
+					g := NewWithT(b)
+
+					ctx, cancel := context.WithCancel(context.Background())
+					defer cancel()
+
+					kine := newKineServer(ctx, b, &kineOptions{
+						backendType: backendType,
+						setup: func(ctx context.Context, tx *sql.Tx) error {
+							return setup(ctx, tx, payload.size, b.N)
+						},
+					})
+
+					kine.ResetMetrics()
+					b.StartTimer()
+					resp, err := kine.client.Get(ctx, "key/", clientv3.WithPrefix())
 
-			ctx, cancel := context.WithCancel(context.Background())
-			defer cancel()
+					g.Expect(err).To(BeNil())
+					g.Expect(resp.Kvs).To(HaveLen((b.N + 1) / 2))
+					kine.ReportMetrics(b)
+				})
 
-			kine := newKineServer(ctx, b, &kineOptions{
-				backendType: backendType,
-				setup: func(db *sql.DB) error {
-					return setupScenario(ctx, db, "key", b.N*2, b.N, b.N)
-				},
+				b.Run("pagination", func(b *testing.B) {
+					b.StopTimer()
+					g := NewWithT(b)
+
+					ctx, cancel := context.WithCancel(context.Background())
+					defer cancel()
+
+					kine := newKineServer(ctx, b, &kineOptions{
+						backendType: backendType,
+						setup: func(ctx context.Context, tx *sql.Tx) error {
+							return setup(ctx, tx, payload.size, b.N)
+						},
+					})
+
+					kine.ResetMetrics()
+					b.StartTimer()
+					nextKey := "key/"
```

Comment:
The logic around `nextKey` is very strange to me, could this be simplified somehow? Also, it looks like the initial `key/` gets lost, is that okay?

Reply:
Yeah we don't really care about the key, but we just want to exercise the database on different ranges (to avoid caching effects), so that we know the "random access" speed of the query.

Regarding the `\x01`, I agree it is weird, but that weirdness *comes from etcd* (see for example here for how [kubernetes uses pagination](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go#L326)). You will notice though that we are using `\x01` instead of `\x00`. The reason for that is a [bug in dqlite protocol](https://dqlite.io/docs/reference/wire-protocol) which is hard to fix without breaking backward compatibility. The dqlite team is already aware of that.

Reply:
Ah this is unfortunate, thanks for the info!